### ISSUE TITLE
timeout not increasing above 34s

### ISSUE THREAD

**Issue by elkimek** [sparrowwallet/sparrow#313]
Hi,

I'm having issues syncing my wallet using electrs 0.9.2 - from Sparrow log I can see there is a timeout which is not increasing above 34s. I had similar issues connecting from my Dojo to electrs and I had to bump the timeout as well to make it working. Note that the timeout from Dojo to electrs was enough with electrs 0.9.0 but it seems to me 0.9.2 needs much more time. I replicated this on two devices and different installations. The only logical solution for my case seems to be increase the timeout above the 34s.

```
2021-11-14 15:04:17,810 WARN [Thread-232] c.s.s.n.TcpTransport [null:-1] No response from server, setting read timeout to 8 secs
2021-11-14 15:04:36,957 WARN [Thread-236] c.s.s.n.TcpTransport [null:-1] No response from server, setting read timeout to 16 secs
2021-11-14 15:04:55,459 WARN [Thread-235] c.s.s.n.TcpTransport [null:-1] No response from server, setting read timeout to 34 secs
2021-11-14 15:06:26,182 WARN [Thread-232] c.s.s.n.TcpTransport [null:-1] No response from server, setting read timeout to 34 secs
2021-11-14 15:07:02,384 WARN [Thread-233] c.s.s.n.TcpTransport [null:-1] No response from server, setting read timeout to 34 secs
2021-11-14 15:07:36,435 WARN [Thread-236] c.s.s.n.TcpTransport [null:-1] No response from server, setting read timeout to 34 secs
2021-11-14 15:08:12,936 WARN [Thread-235] c.s.s.n.TcpTransport [null:-1] No response from server, setting read timeout to 34 secs
2021-11-14 15:08:56,537 WARN [Thread-234] c.s.s.n.TcpTransport [null:-1] No response from server, setting read timeout to 34 secs
2021-11-14 15:09:33,939 WARN [Thread-232] c.s.s.n.TcpTransport [null:-1] No response from server, setting read timeout to 34 secs
2021-11-14 15:10:10,141 WARN [Thread-233] c.s.s.n.TcpTransport [null:-1] No response from server, setting read timeout to 34 secs
2021-11-14 15:10:44,192 WARN [Thread-236] c.s.s.n.TcpTransport [null:-1] No response from server, setting read timeout to 34 secs
2021-11-14 15:11:20,694 WARN [Thread-235] c.s.s.n.TcpTransport [null:-1] No response from server, setting read timeout to 34 secs
2021-11-14 15:12:04,294 WARN [Thread-234] c.s.s.n.TcpTransport [null:-1] No response from server, setting read timeout to 34 secs
2021-11-14 15:12:41,702 WARN [Thread-232] c.s.s.n.TcpTransport [null:-1] No response from server, setting read timeout to 34 secs
2021-11-14 15:13:17,904 WARN [Thread-233] c.s.s.n.TcpTransport [null:-1] No response from server, setting read timeout to 34 secs
2021-11-14 15:13:51,955 WARN [Thread-236] c.s.s.n.TcpTransport [null:-1] No response from server, setting read timeout to 34 secs
2021-11-14 15:14:28,456 WARN [Thread-235] c.s.s.n.TcpTransport [null:-1] No response from server, setting read timeout to 34 secs
```

**Comment by craigraw** [comment]
The problem with increasing timeouts beyond a certain point is that the server is already in trouble if you need to wait half a minute for it to return data from a single request. At some point Sparrow needs to give up and tell the user things haven't worked - otherwise, it appears that Sparrow itself is broken. 

> Note that the timeout from Dojo to electrs was enough with electrs 0.9.0 but it seems to me 0.9.2 needs much more time.

This is strange - v0.9.2 contains several optimizations that should improve performance rather than reduce it. Certainly no request should take this long - I think there is a problem there. Can you add an issue under the Electrs Github issues? @romanz

**Comment by elkimek** [comment]
Upgraded to 0.9.3 and I still have the same issues however I noticed that even when Sparrow timeouts from electrs, electrs is still running and getting the txs. See to following timeline and take note there is an hour difference because of different TZ.

```2021-11-20 15:23:03,137 WARN [Thread-326] c.s.s.n.TcpTransport [null:-1] No response from server, setting read timeout to 8 secs
2021-11-20 15:25:04,091 WARN [Thread-344] c.s.s.n.TcpTransport [null:-1] No response from server, setting read timeout to 8 secs
2021-11-20 15:25:17,069 WARN [Thread-347] c.s.s.n.TcpTransport [null:-1] No response from server, setting read timeout to 16 secs
2021-11-20 15:44:48,665 WARN [Thread-11] c.s.s.n.TcpTransport [null:-1] No response from server, setting read timeout to 8 secs
2021-11-20 15:45:00,366 WARN [Thread-14] c.s.s.n.TcpTransport [null:-1] No response from server, setting read timeout to 16 secs
2021-11-20 15:45:40,450 WARN [Thread-11] c.s.s.n.TcpTransport [null:-1] No response from server, setting read timeout to 34 secs
2021-11-20 15:46:16,651 WARN [Thread-12] c.s.s.n.TcpTransport [null:-1] No response from server, setting read timeout to 34 secs
2021-11-20 15:46:53,154 WARN [Thread-14] c.s.s.n.TcpTransport [null:-1] No response from server, setting read timeout to 34 secs
2021-11-20 15:47:36,756 WARN [Thread-13] c.s.s.n.TcpTransport [null:-1] No response from server, setting read timeout to 34 secs
2021-11-20 15:48:10,809 WARN [Thread-16] c.s.s.n.TcpTransport [null:-1] No response from server, setting read timeout to 34 secs
2021-11-20 15:48:48,211 WARN [Thread-11] c.s.s.n.TcpTransport [null:-1] No response from server, setting read timeout to 34 secs
```

Killed Sparrow here, electrs continues getting the txs

```
[2021-11-20T14:49:31.029Z DEBUG electrs::status] 2 transactions from 2 blocks
[2021-11-20T14:49:31.124Z DEBUG electrs::status] 2 transactions from 2 blocks
[2021-11-20T14:49:31.124Z DEBUG electrs::p2p] loading 1 blocks
[2021-11-20T14:49:31.180Z DEBUG electrs::status] 2 transactions from 2 blocks
[2021-11-20T14:49:31.183Z DEBUG electrs::p2p] loading 1 blocks
[2021-11-20T14:49:31.267Z DEBUG electrs::p2p] loading 1 blocks
[2021-11-20T14:49:31.326Z DEBUG electrs::p2p] loading 1 blocks
[2021-11-20T14:49:31.404Z DEBUG electrs::status] 2 transactions from 2 blocks
[2021-11-20T14:49:31.404Z DEBUG electrs::p2p] loading 1 blocks
[2021-11-20T14:49:31.495Z DEBUG electrs::status] 2 transactions from 2 blocks
[2021-11-20T14:49:31.495Z DEBUG electrs::p2p] loading 1 blocks
```

I'm not sure if I understand the logs correctly but it seems to me that there's either lack of communication between client and server or Sparrow is just impatient and doesn't wait for electrs to finish.

**Comment by romanz** [comment]
Could you please share the following Prometheus statistics:
```
$ curl -s localhost:4224 | grep 'rpc_duration_sum'
$ curl -s localhost:4224 | grep 'rpc_duration_count'
```

**Comment by romanz** [comment]
Also, could you please test how much time does it take to sync the history of the [`xpub6BiVtCpG9fQPxnPmHXG8PhtzQdWC2Su4qWu6XW9tpWFYhxydCLJGrWBJZ5H6qTAHdPQ7pQhtpjiYZVZARo14qHiay2fvrX996oEP42u8wZy` wallet](https://github.com/romanz/electrs/issues/539#issuecomment-940229419)?

**Comment by elkimek** [comment]
```
rpc_duration_sum{method="blockchain.headers.subscribe"} 0.006721908
rpc_duration_sum{method="blockchain.scripthash.get_history"} 944.2643033309994
```

```
rpc_duration_count{method="blockchain.headers.subscribe"} 31
rpc_duration_count{method="blockchain.scripthash.get_history"} 10550
```

> Also, could you please test how much time does it take to sync the history of the [`xpub6BiVtCpG9fQPxnPmHXG8PhtzQdWC2Su4qWu6XW9tpWFYhxydCLJGrWBJZ5H6qTAHdPQ7pQhtpjiYZVZARo14qHiay2fvrX996oEP42u8wZy` wallet](https://github.com/romanz/electrs/issues/539#issuecomment-940229419)?

27 seconds, 108 txs if that's correct (you mention 109 in the link)

**Comment by elkimek** [comment]
I think I identified the issue. Electrs can serve only one wallet at a time, right? Correct me if I'm wrong here @romanz but when I open Sparrow, all four accounts start finding transactions at once.

I tried to import just one account as an Watch Only Wallet and everything went just fine. So it seems to me when I open the wallet the usual way, all four accounts just interrupt each other until Sparrow gives up?

I found out because when I tried to import xpub in Dojo and opened Sparrow, the import process Dojo was running got interrupted.

**Comment by craigraw** [comment]
> Sparrow is just impatient and doesn't wait for electrs to finish.

This is probably true for this case, but it's also possible that certain requests never return, so having a sane timeout is important to handle connectivity errors. 

>  Electrs can serve only one wallet at a time, right?

Electrs can serve many wallets at once - from Electrs' POV, there are no differences between one wallet and another (it just sees things like script hashes and txids as inputs). Sparrow opens one TCP connection to the server, and sends requests to Electrs sequentially in order to receive the right response for the right request. 

> I tried to import just one account as an Watch Only Wallet and everything went just fine.

All Whirlpool users using Sparrow are using multiple accounts (sometimes with many addresses subscribed due to large postmix accounts), so Electrs handling multiple accounts is common. 

Have you observed the the process activity on the Dojo with `top`? Possibly there are other processes or hardware limitations causing performance issues.

**Comment by elkimek** [comment]
I even tried to use the electrs with Sparrow when all other Nodl apps are disabled.

The only difference I'm sure of now is that syncing the postmix account alone works just fine but with the other 3 accounts it just doesn't work. Even from observing electrs logs I can tell that the postmix account is looping in circles until Sparrow gives up.

**Comment by Kixunil** [comment]
> it's also possible that certain requests never return

In practice much less likely than requests taking long time.

> so having a sane timeout is important to handle connectivity errors

A "sane" timeout is a function of complexity of the request. Either Sparrow should make smaller requests or it should increase timeout with the size of the request. Alternatively, if a user can set the timeout it can help users who know to have large wallets.
Or not having a timeout and letting user cancel the request can work - wallets are by their nature interactive.

> Electrs can serve many wallets at once

Yes and no. It has to communicate with Bitcoin Core which is a bottleneck so some requests have to wait for other requests.

> Sparrow opens one TCP connection to the server, and sends requests to Electrs sequentially in order to receive the right response for the right request.

Do the latter requests have increased timeout?

**Comment by craigraw** [comment]
> In practice much less likely than requests taking long time.

It was quite regular in the 0.8.x series that requests didn't return when several hundred addresses were subscribed. I expect this to be much better now, but Tor is still an unreliable communication layer in any event.

> it should increase timeout with the size of the request

Sparrow increases the timeout with the size of the batched request, adding another 50ms for each additional item in a batched request.

> Do the latter requests have increased timeout?

Not sure I follow the question. Sparrow retries requests with an incrementally increased timeout if they fail to return in time, up to a point (34 secs).

**Comment by Kixunil** [comment]
Would it be possible to make timeouts configurable (at least from a config file or command line) so that people can try different values and see what works?

> Not sure I follow the question.

If Sparrow makes a request and then queues another one does the other one have timeout increased by the timeout of the previous?

**Comment by craigraw** [comment]
> Would it be possible to make timeouts configurable (at least from a config file or command line) so that people can try different values and see what works?

It would be possible yes, I'll look into this. Currently it's an array of integers for the timeouts in seconds - I think what makes the most sense is to have  a configurable additional value which can be appended onto the end of the array for a maximum timeout value.

> If Sparrow makes a request and then queues another one does the other one have timeout increased by the timeout of the previous?

Ah! Yes, it does.

**Comment by craigraw** [comment]
Implemented in 8b42399. 

Sparrow's config file now contains the field `maxServerTimeout` which is by default set to 34s. Sparrow will still start with shorter timeouts as it currently does, but if these are breached the timeout is increased until the configured maximum, where it will remain for the duration of the connection.
